### PR TITLE
CIP12 - Memo Requirement through Broadcasts

### PR DESCRIPTION
**Pull request by jdogresorg**
Please review this proposal to enforce a memo requirement on sends for a given address.

### CODE REVIEW

**Comment by jdogresorg** [comment]
updated CIP as requested with `OPTIONS` instead of `MEMO` messages. Also added a milestones section and a bounty amount.

Please review and ACK / NACK :)

File: cip-0012.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+## Changes ##
+
+### Database 
+ 
+1. Create `addresses` table with the following definition:
+```
+CREATE TABLE addresses (
+    address TEXT UNIQUE,    -- address
+    options INTEGER         -- bitwise op flag (1 << 0; 1 << 1)
```
Comment (deweller) [inline cip-0012.md]: I'm not sure what `1 <<0; 1 << 1` means.  Perhaps change this to `-- bitwise address flags (MEMO_REQUIRED is options AND 0x01)`

File: cip-0012.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+1. When validating a send attempt (send.validate):
+   - Check if destination address requires memo
+   - Check if transaction has a memo included
+
+### Broadcasts
+
+1. When parsing a broadcast (broadcast.parse)
+   - Check for `OPTIONS <INTEGER>` as the broadcast text
+   - Verify that address is not LOCKed by checking for previous LOCK broadcast
+   - Create/Update record in `addresses` table and set OPTIONS
```
Comment (deweller) [inline cip-0012.md]: Perhaps add:

- If INTEGER is greater than than the maximum flag used, then the OPTIONS broadcast will be ignored.  As of CIP 12, only `1` or `0` are accepted.

Comment (deweller) [inline cip-0012.md]: Also:
  - The REQUIRE_MEMO flag (bit one) may be turned on or off as many times as desired with successive OPTIONS broadcast messages.

File: cip-0012.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+Enforcing a memo requirement at the protcol level also has the added benefit that it can discourage predatory behavior by exchanges, such as allowing sends without a memo, and ignoring any support requests by customers who made a deposit _without_ a memo. 
+
+## Specification ##
+
+There will be a 'special' broadcast message that is used to set address options; `OPTIONS <INTEGER>`.
+
+
+#### OPTIONS
+```OPTIONS <INTEGER>```
+ - `INTEGER`: a bitwise operation that identifes the desired options
```
Comment (rubensayshi) [inline cip-0012.md]: `identifes` -> `identifies`

File: cip-0012.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  Created: 2017-06-07
+</pre>
+
+## Abstract ##
+
+Any user should be able to mark an address under their control as memo-required. The protocol should track this memo requirement and prevent any sends to the address without a memo.
+
+## Motivation ##
+
+To lessen end user support costs by enforcing a memo requirement at the protcol level.
```
Comment (rubensayshi) [inline cip-0012.md]: `protcol` -> `protocol`

File: cip-0012.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+## Changes ##
+
+### Database 
+ 
+1. Create `addresses` table with the following definition:
+```
+CREATE TABLE addresses (
+    address TEXT UNIQUE,    -- address
+    options INTEGER         -- bitwise address flags (REQUIRE_MEMO is options AND 0x01)
```
Comment (rubensayshi) [inline cip-0012.md]: max of INTEGER is int8, worth mentioning incase anyone ever wants to implement it outside of sqlite context ;)

**Comment by deweller** [comment]
I will "Squash and merge" when I merge the PR which will git rid of the extra commits.

File: cip-0012.md [hunk 6]
```diff
@@ -51,6 +51,9 @@ CREATE TABLE addresses (
 );
  ```
 
+Note: Max INTEGER in sqlite is int8 which allows for a range of -128 to 127
```
Comment (deweller) [inline cip-0012.md]: Actually, this should be an unsigned int8 with values from 0-255.

File: cip-0012.md (hunk 5 above)
Comment (deweller) [inline cip-0012.md]: This is resolved with the additional note below.

**Comment by deweller** [comment]
@jdogresorg - Can you go ahead and change status to Accepted?  We'll just merge and accept this in one shot.

**Comment by deweller** [comment]
Can you please update the readme.md too?

**Comment by jdogresorg** [comment]
@deweller done :)

File: cip-0012.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+To lessen end user support costs by enforcing a memo requirement at the protocol level.
+
+## Rationale ##
+
+CIP9 allows for specifying a memo in a send transaction, which can reduce the FEES cost for exchanges by allowing customers to make deposits directly to an exchange hot-wallet. The data in the 'memo' field is used to associate a transaction with a specific customer, and removes the requirement that each customer have their own deposit address. However, there is still a non-trivial SUPPORT cost associated with tracking down customer deposits _without_ a memo.
+
+By allowing an exchange to mark their address as memo-required, and rejecting sends without a memo, they are able to reduce their support costs.
+
+Enforcing a memo requirement at the protcol level also has the added benefit that it can discourage predatory behavior by exchanges, such as allowing sends without a memo, and ignoring any support requests by customers who made a deposit _without_ a memo.
```
Comment (deweller) [inline cip-0012.md]: protcol => protocol

**Comment by deweller** [comment]
ACK #c3a3e6f

File: cip-0012.md (hunk 5 above)
Comment (rubensayshi) [inline cip-0012.md]: whoops, sorry, 8 byte int, so int64 >_>

**Comment by rubensayshi** [comment]
final change last change (see inline commit), but this PR needs to be rebased / squashed ...

File: cip-0012.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+ 
+1. Create `addresses` table with the following definition:
+```
+CREATE TABLE addresses (
+    address TEXT UNIQUE,    -- address
+    options INTEGER         -- bitwise address flags (REQUIRE_MEMO is options AND 0x01)
+);
+ ```
+
+Note: Max INTEGER in sqlite is unsigned int8 which allows for a range of 0 to 255
```
Comment (deweller) [inline cip-0012.md]: So in regards to Ruben's note above, how about:

Note: The maximum INTEGER stored in sqlite is 64 bits

File: cip-0012.md (hunk 5 above)
Comment (deweller) [inline cip-0012.md]: sqlite stores a variable sized integer based on the magnitude of the number.  64 bits is the maximum.

How about this to be a little more clear
`options INTEGER         -- bitwise address flags (the REQUIRE_MEMO option is represented by the first bit)`

**Comment by deweller** [comment]
No need to rebase and squash at this time.

When I merge the PR, I will use the **Squash and merge** function.  This will clean up all the extra commits in the final repository history.

**Comment by jdogresorg** [comment]
updates made as suggested :)

**Comment by deweller** [comment]
re-ACK #287286b

**Comment by rubensayshi** [comment]
ACK https://github.com/CounterpartyXCP/cips/pull/28/commits/287286bdaf5b0835c92916c520af12c1571f2ac1

**Pull request merged**